### PR TITLE
Enhancements for building licenses for applications

### PR DESCRIPTION
**Pull request by bmcdavid**
added solution parsing support
fallback for failed nuspec requests
added project filter and allowable licenses
added license validation and validation result
added manual information, ability to include / disable project urls

### CODE REVIEW

**Comment by bmcdavid** [comment]
Hi Tom,

I found your excellent project and needed to add some functionality to create license JSON files for my company's application. I figured I would share and hope you find them helpful. Feel free to ask me any questions.

Thanks,
Brad

File: tests/NugetUtility.Tests/ProgramTests.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+
+        [Test]
+        public async Task Main_Should_Error_Empty_Input()
+        {
+            var status = await Program.Main("-j".Split(' '));
+
+            Assert.AreEqual(1, status);
+        }
+
+        [TestCase("-i " + TestSetup.ThisProjectSolutionPath + @" --allowed-license-types ..\..\..\SampleAllowedLicenses.json")]
```
Comment (tomchavakis) [inline tests/NugetUtility.Tests/ProgramTests.cs:31]: TestCase Fails on Linux due to Windows Folder Separator Structure

File: tests/NugetUtility.Tests/ProgramTests.cs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+        [Test]
+        public void Main_Should_Throw_When_InvalidLicenses(string args)
+        {
+            Assert.ThrowsAsync<InvalidLicensesException>(async () => await Program.Main(args.Split(' ')));
+        }
+
+        [Test]
+        public async Task Main_Should_Run_For_This_Project()
+        {
+            var status = await Program.Main(@"-i ..\..\..\".Split(' '));
```
Comment (tomchavakis) [inline tests/NugetUtility.Tests/ProgramTests.cs:41]: same here

File: tests/NugetUtility.Tests/TestSetup.cs [hunk 3]
```diff
@@ -0,0 +1,10 @@
+﻿using NUnit.Framework;
+
+namespace NugetUtility.Tests
+{
+    [SetUpFixture]
+    public class  TestSetup
+    {
+        public const string ThisProjectSolutionPath = @"..\..\..\..\..\NugetUtility.sln";
```
Comment (tomchavakis) [inline tests/NugetUtility.Tests/TestSetup.cs:8]: Change to Path.DirectorySeparatorChar for Windows/Unix-based systems

File: tests/NugetUtility.Tests/ProgramTests.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            Assert.IsTrue(outputFile.Exists, "File doesn't exist");
+            var outputList = Utilties.ReadListFromFile<LibraryInfo>(outputFile.FullName);
+            Assert.IsTrue(outputList.Any(l => l.PackageName == "ASamplePackage"));
+        }
+
+#if WINDOWS
+        [Test]
+        public async Task Main_Should_Run_For_This_Solution_With_Custom_Outfile_FullPath()
+        {
+            const string args = "-i " + TestSetup.ThisProjectSolutionPath + @" -j --outfile c:\temp\custom.json";
```
Comment (tomchavakis) [inline tests/NugetUtility.Tests/ProgramTests.cs:73]: same here

File: tests/NugetUtility.Tests/MethodsTests.cs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+            Assert.IsFalse(result.SelectMany(p => p.Value).Count(p => p.Key == "Onion.SolutionParser.Parser.Standard") > 0);
+        }
+
+        [Test]
+        public async Task GetPackages_AllowedLicenses_Should_Throw_On_Empty()
+        {
+            var methods = new Methods(new PackageOptions
+            {
+                ProjectsFilterOption = @"..\..\..\SampleProjectFilters.json",
+                AllowedLicenseTypesOption = @"..\..\..\SampleAllowedLicenses.json",
```
Comment (tomchavakis) [inline tests/NugetUtility.Tests/MethodsTests.cs:99]: same here

File: tests/NugetUtility.Tests/MethodsTests.cs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+    [TestFixture]
+    public class MethodTests
+    {
+        private string _projectPath;
+        private Methods _methods;
+
+        [SetUp]
+        public void Setup()
+        {
+            _projectPath = @"..\..\..\";
```
Comment (tomchavakis) [inline tests/NugetUtility.Tests/MethodsTests.cs:18]: same here

**Comment by tomchavakis** [comment]
Hi @bmcdavid ,

I would like to thank you for your awesome contribution !!!
I use Linux and VSCode and I had difficulty running the tests due to Windows path setup, but after altering them to the Unix-based path, runs successfully. 

Could you please change the README usage in order to be updated?
Would you mind adding your username as a contributor to this library? 

Thank's in advance

**Comment by bmcdavid** [comment]
Hi @tomchavakis,

I've fixed the path issue to work for all environments and updated the readme. Also feel free to add my username as a contributor :).

Thanks,
Brad

**Comment by bmcdavid** [comment]
@tomchavakis Any chance you can make a new NuGet package on NuGet.org with these changes?

Thanks in advance!

**Comment by tomchavakis** [comment]
Hi @bmcdavid,
I have released the new version of nuget-license project.
You can find it at releases and as a Nuget package.
https://www.nuget.org/packages/dotnet-project-licenses/2.0.0

Thanks again for your contribution.

**Pull request merged**